sql/src/main/java/io/crate/jobs/ExecutionSubContextBase.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.jobs;
+
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.CancellationException;
+
+public abstract class ExecutionSubContextBase implements ExecutionSubContext, ExecutionState {
+
+    protected static final ESLogger LOGGER = Loggers.getLogger(ExecutionSubContextBase.class);
+
+    protected final SubExecutionContextFuture future;
+    protected JobExecutionContext context;
+    private final int id;
+
+    // TODO: remove this flag when ExecutionState is obsolete
+    private volatile boolean isKilled = false;
+
+
+    protected ExecutionSubContextBase(int id) {
+        this.id = id;
+        future = new SubExecutionContextFuture(this);
+    }
+
+    @Override
+    public void jobExecutionConttext(JobExecutionContext context) {
+        this.context = context;
+    }
+
+    public int id() {
+        return id;
+    }
+
+    protected void innerPrepare() {
+    }
+
+    @Override
+    public void prepare() {
+        if (!future.closed()) {
+            LOGGER.trace("preparing {}: {}", this, id);
+            try {
+                innerPrepare();
+            } catch (Throwable t) {
+                close(t);
+            }
+        }
+    }
+
+    protected void innerStart() {
+    }
+
+    @Override
+    public void start() {
```

Comment:
consider making this final so that sub-classes can't overwrite it but use innerStart. Same for prepare, etc.